File: HairPlus.Web/app/controllers/expenseController.js
(function (module) {

    var expenseController = function ($scope, toastr, expenseService) {

        $scope.expenses = [];

        $scope.pagingInfo = {
            page: 1,
            itemsPerPage: 50,
            sortBy: 'amount',
            reverse: false,
            search: '',
            totalItems: 0
        };

        $scope.deleteExpense = function (expense) {
            if (confirm("Are you sure..?")) {
                expenseService.deleteExpense(expense.expenseId).then(function (data) {
                    $scope.expenses.splice($scope.expenses.indexOf(expense), 1);
                    toastr.success("Expense deleted successfully")
                }, function (err) {
                    console.log(err);
                    toastr.error("unable to delete expense")
                })
            }
        }

        var onExpenses = function (data) {
            $scope.expenses = data.data.data;
            $scope.pagingInfo.totalItems = data.data.count;
        }

        var onExpensesError = function (err) {
            console.log(err);
            toastr.error('unable to load expenses');
        }

        var init = function () {
            expenseService.getExpenses($scope.pagingInfo).then(onExpenses, onExpensesError);
        }

        init();

    }

    expenseController.$inject = ["$scope", "toastr", "expenseService"];
    module.controller("expenseController", expenseController);

    /************* Add Or Update expense Controller ****************/

    var addUpdateExpenseController = function ($scope, $state, $stateParams, toastr, expenseService) {

        $scope.model = {
            expenseId: 0,
            amount: 0,
            description: "",
            expenseDateTime: new Date()
        };
        $scope.isNewExpense = true;

        $scope.isOpen = false;

        $scope.openCalendar = function (e) {
            e.preventDefault();
            e.stopPropagation();
            $scope.isOpen = true;
        };

        $scope.saveExpense = function () {
            expenseService.addExpense($scope.model).then(function () {
                toastr.success("expense added successfully");
                $state.go("expenses");
            }, function (err) {
                console.log(err);
                toastr.error("unable to save expense at this time");
            })
        }

        $scope.updateExpense = function () {
            expenseService.updateExpense($scope.model).then(function () {
                toastr.success("expense updated successfully");
                $state.go("expenses");
            }, function (err) {
                console.log(err);
                toastr.error("unable to update expense at this time");
            })
        }

        var onExpense = function (data) {
            data = data.data;
            $scope.model.amount = data.amount;
            $scope.model.description = data.description;
            $scope.model.expenseDateTime = data.expenseDateTime;
        }

        var init = function () {
            if ($state.current.name != "add-expense") {
                $scope.isNewExpense = false;
                var expenseId = $stateParams.eId;
                $scope.model.expenseId = expenseId;
                expenseService.getExpense(expenseId).then(onExpense, function (err) {
                    console.log(err);
                    toastr.error("unable to fetch expense data");
                })
            }
        }

        init();
    }

    addUpdateExpenseController.$inject = ["$scope", "$state", "$stateParams", "toastr", "expenseService"];
    module.controller("addUpdateExpenseController", addUpdateExpenseController);


}(angular.module("HairPlus.controllers")))

File: HairPlus.Web/Controllers/ExpenseController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Web.Http;
using Microsoft.AspNet.Identity;
using System.Data.Entity;
using System.Threading.Tasks;
using HairPlus.Contract;
using HairPlus.Web.Models;
using System.Linq.Dynamic;

namespace HairPlus.Web.Controllers
{
    public class ExpenseController : BaseController
    {
        public ExpenseController(IUow uow)
        {
            _Uow = uow;
        }

        [HttpGet]
        [ActionName("GetAll")]
        public async Task<IHttpActionResult> GetAll(int page = 1, int itemsPerPage = 20, string sortBy = "amount", bool reverse = false, string search = null)
        {
            try
            {
                var expenseList = new List<ExpenseListViewModel>();

                var expenses = _Uow._Expense.GetAll(x => x.Active == true);

                // searching
                if (!string.IsNullOrWhiteSpace(search))
                {
                    search = search.ToLower();
                    expenses = expenses.Where(x =>
                        x.Description.ToLower().Contains(search));
                }

                // sorting (done with the System.Linq.Dynamic library available on NuGet)
                expenses = expenses.OrderBy(sortBy + (reverse ? " descending" : ""));

                var totalExpenses = await expenses.CountAsync();

                // paging
                var expensesPaged = await expenses.Skip((page - 1) * itemsPerPage).Take(itemsPerPage).ToListAsync();

                foreach (var item in expensesPaged)
                {
                    ExpenseListViewModel model = new ExpenseListViewModel();

                    model.ExpenseDateTime = item.CreatedOn;
                    model.Amount = item.Amount;
                    model.Description = item.Description;
                    model.ExpenseId = item.Id;

                    expenseList.Add(model);
                }

                // json result
                var json = new
                {
                    count = expensesPaged,
                    data = expenseList,
                };

                return Ok(json);
            }
            catch (Exception ex)
            {
                return InternalServerError(ex);
            }
        }

        [HttpGet]
        [ActionName("GetExpense")]
        public async Task<IHttpActionResult> GetExpense(int id)
        {
            try
            {
                var expense = await _Uow._Expense.GetAsync(x => x.Id == id && x.Active == true);

                if (expense == null)
                {
                    return NotFound();
                }

                var model = new ExpenseListViewModel();

                model.Amount = expense.Amount;
                model.Description = expense.Description;
                model.ExpenseId = expense.Id;
                model.ExpenseDateTime = expense.CreatedOn;

                return Ok(model);
            }
            catch (Exception ex)
            {
                return InternalServerError(ex);
            }
        }

        [HttpPost]
        [ActionName("AddExpense")]
        public async Task<IHttpActionResult> AddExpense(ExpenseListViewModel model)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return BadRequest("Data not valid. Please enter valid data and try again");
                }

                var expense = new EF.Expense
                {
                    Active = true,
                    Description = model.Description,
                    Amount = model.Amount,
                    CreatedBy = User.Identity.GetUserId(),
                    CreatedOn = model.ExpenseDateTime,
                };

                _Uow._Expense.Add(expense);
                await _Uow.CommitAsync();
                return Ok();
            }
            catch (Exception ex)
            {
                return InternalServerError(ex);
            }
        }

        [HttpPost]
        [ActionName("UpdateExpense")]
        public async Task<IHttpActionResult> UpdateExpense(ExpenseListViewModel model)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return BadRequest("Data not valid. Please enter valid data and try again");
                }

                var expense = await _Uow._Expense.GetAsync(x => x.Id == model.ExpenseId && x.Active == true);

                if (expense == null)
                {
                    return NotFound();
                }
                expense.Amount = model.Amount;
                expense.Description = model.Description;
                expense.UpdatedBy = User.Identity.GetUserId();
                expense.UpdatedOn = DateTime.Now;

                _Uow._Expense.Update(expense);
                await _Uow.CommitAsync();
                return Ok();
            }
            catch (Exception ex)
            {
                return InternalServerError(ex);
            }
        }

        [HttpDelete]
        [ActionName("DeleteExpense")]
        public async Task<IHttpActionResult> DeleteExpense(int id)
        {
            try
            {
                var expense = await _Uow._Expense.GetAsync(x => x.Id == id && x.Active == true);

                if (expense == null)
                {
                    return NotFound();
                }

                expense.Active = false;
                _Uow._Expense.Update(expense);
                await _Uow.CommitAsync();
                return Ok();
            }
            catch (Exception ex)
            {
                return InternalServerError(ex);
            }
        }
    }
}
